### PR TITLE
Encapsulate the buffers within PixelBuf

### PR DESCRIPTION
**Pull request by tannewt**
Also changes MicroPython so that native methods get a subclass instance instead of the native object directly. This may break things.

### CODE REVIEW

File: py/objtype.c [hunk 1]
```diff
@@ -187,14 +187,7 @@ STATIC void mp_obj_class_lookup(struct class_lookup_data  *lookup, const mp_obj_
...
                     mp_obj_instance_t *obj = lookup->obj;
-                    mp_obj_t obj_obj;
-                    if (obj != NULL && mp_obj_is_native_type(type) && type != &mp_type_object /* object is not a real type */) {
-                        // If we're dealing with native base class, then it applies to native sub-object
-                        obj_obj = obj->subobj[0];
-                    } else {
-                        obj_obj = MP_OBJ_FROM_PTR(obj);
-                    }
-                    mp_convert_member_lookup(obj_obj, type, elem->value, lookup->dest);
+                    mp_convert_member_lookup(MP_OBJ_FROM_PTR(obj), type, elem->value, lookup->dest);
```
Comment (rhooper) [inline py/objtype.c]: If this has no unexpected side effects this is a nice little cleanup!

File: shared-module/_pixelbuf/PixelBuf.c [hunk 2]
```diff
@@ -26,95 +26,256 @@
...
+    size_t pixel_len = self->pixel_count * self->bytes_per_pixel;
+    self->transmit_buffer_obj = mp_obj_new_bytes_of_zeros(header_len + pixel_len + trailer_len);
+    mp_obj_str_t *o = MP_OBJ_TO_PTR(self->transmit_buffer_obj);
+
+    // Abuse the bytes object a bit by mutating it's data by dropping the const. If the user's
+    // Python code holds onto it, they'll find out that it changes. At least this way it isn't
+    // mutable by the code itself.
+    uint8_t* transmit_buffer = (uint8_t*) o->data;
+    memcpy(transmit_buffer, header, header_len);
+    memcpy(transmit_buffer + header_len + pixel_len, trailer, trailer_len);
```
Comment (rhooper) [inline shared-module/_pixelbuf/PixelBuf.c]: Correct me if I'm wrong, but doesn't this make the header and trailer fixed at PixelBuf construction?   While this is fine for Dotstar, what if some other protocol needs to manipulate the header or trailer?  (Not that this is necessarily a bad thing).

File: shared-module/_pixelbuf/PixelBuf.c [hunk 3]
```diff
@@ -26,95 +26,256 @@
...
+    self->post_brightness_buffer = transmit_buffer + header_len;
+
+    if (self->byteorder.is_dotstar) {
```
Comment (rhooper) [inline shared-module/_pixelbuf/PixelBuf.c]: While we're abstracting things, perhaps we should move this kind of initialization to another function or method.  It's always bothered me that the base PixelBuf has to know the details of DotStar, rather than having a subclass or implementation that provides this.

File: shared-module/_pixelbuf/PixelBuf.c [hunk 4]
```diff
@@ -26,95 +26,256 @@
...
+mp_float_t common_hal__pixelbuf_pixelbuf_get_brightness(mp_obj_t self_in) {
+    pixelbuf_pixelbuf_obj_t* self = native_pixelbuf(self_in);
+    return self->brightness;
+}
+
+void common_hal__pixelbuf_pixelbuf_set_brightness(mp_obj_t self_in, mp_float_t brightness) {
+    pixelbuf_pixelbuf_obj_t* self = native_pixelbuf(self_in);
+    self->brightness = brightness;
+    size_t pixel_len = self->pixel_count * self->bytes_per_pixel;
+    if (self->pre_brightness_buffer == NULL) {
```
Comment (rhooper) [inline shared-module/_pixelbuf/PixelBuf.c]: Couldn't this malloc and copy be avoided if brightness >= 1.0?
If this malloc fails, will the error that happens be easy to understand?

File: py/objtype.c (hunk 1 above)
Comment (tannewt) [reply py/objtype.c]: I expect side effects so they won't be unexpected. :-) We can hunt them down as they come up.

File: shared-module/_pixelbuf/PixelBuf.c (hunk 2 above)
Comment (tannewt) [reply shared-module/_pixelbuf/PixelBuf.c]: It does fix it. We can deal with other protocols when we find one that needs it.

File: shared-module/_pixelbuf/PixelBuf.c (hunk 3 above)
Comment (tannewt) [reply shared-module/_pixelbuf/PixelBuf.c]: Ya, I agree this is weird. I don't really want to tackle it now though. My main concern was the buffer management.

File: shared-module/_pixelbuf/PixelBuf.c (hunk 4 above)
Comment (tannewt) [reply shared-module/_pixelbuf/PixelBuf.c]: Yup, in fact I need to add this. Otherwise we'll always allocate it because the constructor calls it. Will do now. :-)

Comment (tannewt) [reply shared-module/_pixelbuf/PixelBuf.c]: Ok, this is updated. I confirmed with `gc.mem_frees` and a large number of pixels that the allocation happens once brightness is set to something besides 1.0.

File: py/objtype.c (hunk 1 above)
Comment (rhooper) [reply py/objtype.c]: The unit tests pass... 🤷‍♂ Probably some edge cases in inheritance again.  I wish we could exercise all the examples in learn!

**Comment by tannewt** [comment]
Ok, this is ready for another look. I got some space back by tweaking a couple error messages.

File: ports/atmel-samd/common-hal/busio/UART.c [hunk 5]
```diff
@@ -150,7 +150,7 @@ void common_hal_busio_uart_construct(busio_uart_obj_t *self,
         self->buffer = (uint8_t *) gc_alloc(self->buffer_length * sizeof(uint8_t), false, true);
```
Comment (jepler) [inline ports/atmel-samd/common-hal/busio/UART.c]: Why not use m_malloc here?  You'd get the error message for free.

File: py/objlist.c [hunk 6]
```diff
@@ -46,6 +46,7 @@ STATIC mp_obj_t list_pop(size_t n_args, const mp_obj_t *args);
 
 STATIC void list_print(const mp_print_t *print, mp_obj_t o_in, mp_print_kind_t kind) {
     mp_obj_list_t *o = MP_OBJ_TO_PTR(o_in);
+    //mp_obj_list_t *o = mp_instance_cast_to_native_base(o_in, &mp_type_list);
```
Comment (jepler) [inline py/objlist.c]: Not sure why this line is added.

File: ports/atmel-samd/common-hal/busio/UART.c (hunk 5 above)
Comment (tannewt) [reply ports/atmel-samd/common-hal/busio/UART.c]: So that we can deinit the UART before throwing the error.

File: py/objlist.c (hunk 6 above)
Comment (tannewt) [reply py/objlist.c]: Ah, I thought I might need it. Will remove it.

**Comment by tannewt** [comment]
Removed and merged in latest master.
